cmd/template_cmd.go

```diff
@@ -343,3 +367,114 @@ func TemplateServersList(c *cli.Context) error {
+	cookbookVersions := make([]*types.CookbookVersion, 0)
+	for _, cbvIn := range strings.Split(cbvsIn, ",") {
+		values := regexp.MustCompile(`(.*?)(~>|=|>=|<=|>|<|:)(\w.*)`).FindStringSubmatch(cbvIn)
+		if len(values) == 0 {
+			return nil, fmt.Errorf("invalid input cookbook version format %s", cbvIn)
+		}
+		name, operator, version := values[1], values[2], values[3]
+		if _, found := result[name]; found {
+			return nil, fmt.Errorf("detected duplicated cookbook version name: %s", name)
+		}
+
+		// uploaded. It requires to map adequate version_id
+		if operator == ":" {
+			if len(cookbookVersions) == 0 {
+				// data is loaded only once
+				svc, formatter := WireUpCookbookVersion(c)
+				cbvs, err := svc.GetCookbookVersionList()
+				if err != nil {
+					formatter.PrintFatal("cannot receive uploaded cookbook versions data", err)
+				}
+				cookbookVersions = cbvs
+			}
+			for _, cbv := range cookbookVersions {
+				if name == cbv.Name && version == cbv.Version {
+					result[name] = map[string]interface{}{"version_id": cbv.ID}
+				}
+			}
+			// name does not match the available uploaded
+			if _, found := result[name]; !found {
+				return nil, fmt.Errorf("invalid uploaded cookbook version name: %s", name)
+			}
+		} else {
+			//supermarket
+			result[name] = map[string]interface{}{"version": operator + " " + version} // at any case, it should leave a space between operator and version
+		}
+	}
+	return result, nil
+}
+
+// convertFlagParamsToParameterValues returns the json representation for the given friendly input format of parameter-values assignation
+func convertFlagParamsToParameterValues(c *cli.Context, cbIn string) (map[string]string, error) {
+	result := map[string]string{}
+	for _, cb := range strings.Split(cbIn, ",") {
+		values := regexp.MustCompile(`(.*?)(:)(\w.*)`).FindStringSubmatch(cb)
+		if len(values) == 0 {
+			return nil, fmt.Errorf("invalid input parameter values format %s", cb)
+		}
+		name, _, value := values[1], values[2], values[3]
+		if _, found := result[name]; found {
+			return nil, fmt.Errorf("detected duplicated parameter name: %s", name)
+		}
+		result[name] = value
+	}
+	return result, nil
+}
+
+// convertFlagParamsToConfigurationAttributesFromFile returns the json representation of configuration attributes taken from the input file or STDIN
+func convertFlagParamsToConfigurationAttributesFromFile(c *cli.Context, casIn string) (map[string]interface{}, error) {
+	var content map[string]interface{}
+	if casIn == "-" {
+		// read from STDIN
+		logrus.Info("Please, write configuration parameters json formatted:")
+		buff := ""
+		scanner := bufio.NewScanner(os.Stdin)
+		for scanner.Scan() {
+			line := scanner.Text()
+			if len(line) == 0 {
+				logrus.Info("Processing...")
+				break
+			}
+			buff += line
+		}
+		if err := scanner.Err(); err != nil {
+			return nil, fmt.Errorf("reading standard input: %s", err)
+		}
+
+		if err := json.Unmarshal([]byte(buff), &content); err != nil {
```

Comment:
Both os.Open and STDIN return/are io.Reader. Instead of reading content to memory in two different ways, use a [json Decoder](https://godoc.org/encoding/json#Decoder)

Reply:
Done